fuel/app/classes/materia/api/v1.php

```diff
@@ -68,6 +68,48 @@ static public function widget_instance_delete($inst_id)
 		return $inst->db_remove();
 	}
 
+	/**
+	 * @return object, contains properties indicating whether the current
+	 * user can edit the widget and a message object describing why, if not
+	 */
+	static public function widget_instance_edit_perms_verify($inst_id)
+	{
+		$response = new \stdClass();
+		$response->msg = null;
+		$response->is_locked = true;
+		$response->can_publish = false;
+
+		$locked_by = \Cache::easy_get('instance-lock.'.$inst_id);
+
+		if ( ! Util_Validator::is_valid_hash($inst_id)) $response->msg = Msg::invalid_input($inst_id);
+		else if (\Service_User::verify_session() !== true) $response->msg = Msg::no_login();
+		else if ( ! static::has_perms_to_inst($inst_id, [Perm::FULL])) $response->msg = Msg::no_perm();
+		else if ( ! ($inst = Widget_Instance_Manager::get($inst_id))) $response->msg = new Msg(Msg::ERROR, 'Widget instance does not exist.');
+
+		//msg property only set if something went wrong
+		if ( ! $response->msg)
+		{
+			$response->is_locked = ! Widget_Instance_Manager::locked_by_current_user($inst_id);
+			$response->can_publish = $inst->widget->publishable_by(\Model_User::find_current_id());
+		}
+
+		return $response;
+	}
+
+	/**
+	 * @return bool, true if the current user can publish the given widget instance, false otherwise.
+	 */
+	static public function widget_publish_perms_verify($widget_id)
+	{
+		if (\Service_User::verify_session() !== true) return Msg::no_login();
+		if ( ! Util_Validator::is_pos_int($widget_id)) return Msg::invalid_input($widget_id);
+
+		$widget = new Widget();
+		if ( $widget->get($widget_id) == false) return Msg::invalid_input('Invalid widget type');
```

Comment:
This error message seems wrong - this error would pop up if the widget instance wasn't found right?

Reply:
Nah, this one specifically is checking to see if a widget (non-instance) exists - hence 'widget type'. Not really sure about 'widget type' as the vocabulary, but 'type' is probably still a better word than 'template'.